File: 2022/06.ts
// https://adventofcode.com/2022/day/6

export class Solver {
  private _input: string[];

  constructor(rawInput: string) {
    this._input = rawInput.split("");
  }

  private _findUniqueCharacterSet(length: number): number {
    let recentUniqueChars = [];

    for (let i = 0; i < this._input.length; i++) {
      const char = this._input[i];
      const matchIndex = recentUniqueChars.indexOf(char);

      if (matchIndex !== -1) {
        // If this char is already in the list,
        // remove the chars up to the match and add the new char
        recentUniqueChars.splice(0, matchIndex + 1);
        recentUniqueChars.push(char);
      } else {
        // If not, just add the char to the list
        recentUniqueChars.push(char);

        if (recentUniqueChars.length === length) {
          return i + 1;
        }
      }
    }
  }

  public part1(): number {
    return this._findUniqueCharacterSet(4);
  }

  public part2(): number {
    return this._findUniqueCharacterSet(14);
  }
}

File: src/bin/06.rs
// https://adventofcode.com/2022/day/6

use std::fs::read_to_string;

fn find_unique_set(input: &str, length: u16) -> u16 {
    let mut recent_unique_chars: Vec<char> = vec![];

    for (i, char) in input.chars().enumerate() {
        let char_position = recent_unique_chars.iter().position(|c| c == &char);

        match char_position {
            Some(position) => {
                // If this char is already in the list, remove
                // the chars up to the match and add the new char
                recent_unique_chars.drain(..position + 1);
                recent_unique_chars.push(char);
            }
            None => {
                // If not, just add the new char to the list
                recent_unique_chars.push(char);

                if recent_unique_chars.len() as u16 == length {
                    return i as u16 + 1;
                }
            }
        }
    }

    0
}

fn part1(input: &str) -> u16 {
    find_unique_set(input, 4)
}

fn part2(input: &str) -> u16 {
    find_unique_set(input, 14)
}

fn main() {
    let file_path = "./input/2022/06-full.txt";

    let input = read_to_string(file_path).expect("Can't read file");
    let input = input.trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}
